src/index.js

```diff
@@ -0,0 +1,59 @@
+const AMQP = require('simple-amqplib');
+const CustomMigrationSource = require('./migration-source');
+const debug = require('debug')('amqp-log-to-pg');
+const onMessage = require('./on-message');
+const ow = require('ow');
+
+const validateConfig = config => {
+  ow(config, ow.object.partialShape({
+    amqp: ow.object.exactShape({
+      url: ow.string.url,
+      exchange: ow.string.nonEmpty,
+      queue: ow.object.exactShape({
+        name: ow.string.nonEmpty,
+        routingKey: ow.string.nonEmpty
+      })
+    }),
+    db: ow.string.url,
+    tableName: ow.string.nonEmpty
+  }));
+};
+
+let amqp;
+let knex;
+let tableName;
+async function init (config) {
+  validateConfig(config);
+
+  tableName = config.tableName;
+  knex = require('knex')(config.db);
+  amqp = new AMQP(config.amqp);
+  debug('running migrations...');
+  await knex.migrate.latest({ migrationSource: new CustomMigrationSource(config.tableName) });
+  debug('connecting to AMQP...');
+  await amqp.connect();
+  debug('ready...');
+}
+
+function consume () {
+  amqp.consume(onMessage(knex, tableName));
+}
+
+async function main () {
+  const config = require('../config');
+  await init(config);
+  consume();
+}
+
+async function shutdown () {
+  debug('shutting down amqp...');
+  await amqp.close();
+  debug('shutting down knex...');
+  await knex.destroy();
+}
+
+if (require.main === module) {
```

Comment:
I get why you are doing this but I also wonder if it wouldn't be better to just make this 2 different files?

```js
// index.js
const config = require('../config');
const { init, consume } = require('./service');

async function main () {
  await init (config);
  consume();
}

main().catch(console.error);

// service.js
module.exports = { init, consume, shutdown };
```

or better yet break the functions into individual functions that can be chained together?
```js
// init.js
async function init (config) {
  // ...
  return { amqp, knex, tableName };
}

// consume.js
function consume (state) {
  const { amqp, knex, tableName } = state;
  amqp.consume(onMessage(knex, tableName));
}

// shutdown.js
async function shutdown (state) {
  const { amqp, knex } = state;
  await amqp.close();
  await knex.destroy();
}
```

Reply:
https://noblesamurai.slack.com/archives/C34988C7P/p1599524947009300?thread_ts=1599521439.008500&cid=C34988C7P

Reply:
```js
const config = require('../config');
async function main () {
  const state = await init(config);
  consume(state);
}
main().catch(console.error);
```